File: Best Time to Buy and Sell Stock.java
class Solution 
{
    public int maxProfit(int[] prices) 
    {
        int profit = 0;
        int buyPrice = prices[0];
        
        // Each iteration we check to see if the current day is the day to buy.
        // If it is NOT the day to buy, we check to see if it is the day to sell to make the most profit. 
        for(int i=1;i<prices.length;i++)
        {
            // if statement is reponsible for finding the price to buy.
            if(prices[i] < buyPrice)    buyPrice = prices[i];
            
            // check if selling on the current day will yield a better profit margin.
             profit = Math.max(profit, prices[i] - buyPrice);
        }
        
        return profit;
    }
}

Input: prices = [7,1,5,3,6,4]
Output: 5
Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.

File: Best Time to Buy and Sell Stock.py
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
  
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        buy = prices[0]
        
        for price in prices:
            if price < buy:
                buy = price
            else:
                max_profit = max(max_profit, price - buy)
        
        return max_profit
-----------------------------------------------------------------------------------------------------------------------------------------------------------  
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if len(prices) == 1:
            return 0
        
        # Assume you always buy on day 1. The reason being you need to buy before you sell.
        buy_in = prices[0]
        current_profit, profit = 0, 0
        
        # Go through each day to see if the current price is less than the price we bought in
        # If it's less, we buy on that day instead.
        # If it's more, we check how much we can profit from that day, and compare with the max profit we currently have.
        for price in prices:
            if price < buy_in:
                buy_in = price
            else:
                current_profit = price - buy_in
            
            profit = max(current_profit, profit)
        
        return profit
